fix(insights): HogQL calculation of saved legacy insights v2

Twixes opened the pull request:
## Changes

Revert of #21590, with ~~one change~~ two changes added:
1. insight serialization will now only use cached results, to match the behavior of the legacy system, preventing ClickHouse overload (5c295c80e65df00e645ddefdf0f2e5661c096cbc)
2. dashboard filters are now taken into account, with proper validation

Twixes committed:
fix(insights): HogQL calculation of saved legacy insights v2

This reverts commit a6314c6bb7603e3d916f38be7a7e845383c95c75.

Twixes committed:
Only use cached results in `process_query` for insight serializer

webjunkie commented on posthog/hogql_queries/query_runner.py:
```diff
@@ -57,6 +58,15 @@
 
 
+class RecalculationMode(Enum):
+    REQUEST = auto()
+    """Always recalculate, except if a VERY recent result is present in cache."""
+    IF_STALE = auto()
+    """Use cache, unless the results are stale or missing."""
+    NEVER = auto()
+    """Only use cache."""
```
Let's be absolutely clear:

```suggestion
    ALWAYS_EXCEPT_RECENT = auto()
    """Always recalculate, except if a VERY recent result is present in cache."""
    USE_CACHE_UNLESS_STALE = auto()
    """Use cache, unless the results are stale or missing."""
    ONLY_USE_CACHE = auto()
    """Only use cache."""
```

Twixes replied:
Oh yeah, this _was_ pretty vague. Renamed somewhat along the lines of your suggestion!

Twixes committed:
Fix type of results

Twixes committed:
Rename `RecalculationMode` to `ExecutionMode`

Twixes committed:
Fix typing more

Twixes committed:
Properly support dashboard filters

Twixes committed:
Hacky fix for schema.py

Twixes committed:
Don't test legacy `generate_insight_cache_key` with `query`

Twixes committed:
Fix importing & typing

Twixes committed:
Fix typo

Twixes committed:
Update test_query_runner.py

Twixes committed:
Account for property filter groups in dashboard filters

Twixes committed:
Do return stale result in CACHE_ONLY case

Twixes committed:
Fix `execute_hogql_query` espionage

Wow, this was a pain to figure out, only was an issue in CI, because the trigger was `TestCohort::test_creating_update_and_calculating_with_new_cohort_query` running prior to `TestInsight:: test_insight_refreshing_query` – had to use trial and error.

Twixes commented on package.json:
```diff
@@ -34,7 +34,8 @@
         "schema:build:json": "ts-node bin/build-schema.mjs && prettier --write frontend/src/queries/schema.json",
-        "schema:build:python": "datamodel-codegen --class-name='SchemaRoot' --collapse-root-models --target-python-version 3.10 --disable-timestamp --use-one-literal-as-default --use-default --use-default-kwarg --use-subclass-enum --input frontend/src/queries/schema.json --input-file-type jsonschema --output posthog/schema.py --output-model-type pydantic_v2.BaseModel && ruff format posthog/schema ...
+        "schema:build:python": "datamodel-codegen --class-name='SchemaRoot' --collapse-root-models --target-python-version 3.10 --disable-timestamp --use-one-literal-as-default --use-default --use-default-kwarg --use-subclass-enum --input frontend/src/queries/schema.json --input-file-type jsonschema --output posthog/schema.py --output-model-type pydantic_v2.BaseModel && ruff format posthog/schema ...
+        "schema:build:python:fix-up-enum": "sed -i '' -e 's/Optional\\[PropertyOperator\\] = \\(\"[A-Za-z_]*\"\\)/Optional[PropertyOperator] = PropertyOperator(\\1)/g' posthog/schema.py # THIS SED INVOCATION IS A MASSIVE HACK - remove when datamodel-codegen properly supports enums",
```
Context: We need `EventPropertyFilter.operator` to have a default value, to support the historic assumption that "exact" is the default operator. But that default value needs to be an enum member (`PropertyOperator.exact`), which `datamodel-codegen` doesn't support

Twixes replied:
Yeah, that's definitely a good idea at this point! Done

Twixes committed:
Fix typing even more

Twixes committed:
Don't require `pnpm` for `schema:build:python`

Matters in CI.

Twixes committed:
Merge branch 'master' into hogql-calculation-bis

Twixes committed:
Fix `schema:build:python`

webjunkie commented on posthog/api/test/test_insight.py:
```diff
@@ -1124,6 +1131,138 @@ def test_insight_refreshing(self, spy_update_insight_cache) -> None:
+                ],
+            )
+            self.assertEqual(response["last_refresh"], "2012-01-16T05:01:34Z")
+            self.assertEqual(response["last_modified_at"], "2012-01-15T04:01:34Z")  # did not change
```
Wait, does that mean it didn't change from the non-dashboard one? Shouldn't it be different? Or it didn't change since creation? In the latter case we might adjust the case to make it clear we get a different not-change for dashboards 🙈

Twixes committed:
Fix sed usage

webjunkie commented on posthog/hogql_queries/query_runner.py:
```diff
@@ -313,8 +335,12 @@ def run(self, refresh_requested: Optional[bool] = None) -> CachedQueryResponse:
                 else:
                     QUERY_CACHE_HIT_COUNTER.labels(team_id=self.team.pk, cache_hit="stale").inc()
+                    if execution_mode == ExecutionMode.CACHE_ONLY:
+                        return cached_response
```
Do we need to set `is_cached = True` here as well? 🤔

Twixes replied:
Hmm, yes, reworked and hopefully clarified this code overall

webjunkie reviewed (approved):
Seems alright in general. Some notes though.

Another thing I'm wondering for later:
If I go to an insight to view it, and it has been updated "1 day ago". I can click refresh there and it'll make just the request to `/query` and then tell me "Computed a few seconds ago". But since we don't store this in the insight, upon refresh it'll be a day old again.
Not sure how to tie this in together... we probably cannot be sure the query is exactly the umodified insight and such 😩

Twixes committed:
Move `schema:build:python` to a bash script

Twixes committed:
Validate cache properly

Clarifies the `cached_response.is_cached = True` situation.

Twixes committed:
Merge branch 'master' into hogql-calculation-bis

Twixes committed:
Fix Python formatting

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

Twixes committed:
Update mypy-baseline.txt

Twixes committed:
Add test to ensure /query/ and /inisghts/ use the same cache

Twixes commented:
> But since we don't store this in the insight, upon refresh it'll be a day old again.

This is exactly what should be addressed here. Just added a test to make sure the two endpoints indeed use the same cache, unlike previously.
